Add unit test for Node controller

cadenmarchese opened the pull request:
### Which issue this PR addresses:

https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/10227470/

<!--
Please include a link to the ADO work item as well as any GitHub issues.

Usage: `Fixes #<GitHub issue number>`, or `Fixes (paste link of issue)`.
-->

### What this PR does / why we need it:

Adds some coverage to the Node controller to ensure we're adding and reading our annotations properly. Also adds a const.go file for the node package since it's standard practice among other operator controllers.
<!--
Include a brief summary of what the PR is intended to accomplish and how the PR
does it. (2-3 sentences)
-->

### Test plan for issue:

`make test-go`
`go test -v`
<!--
How did you test that this PR works?

- Are there unit tests?
- Are there integration/e2e tests?
- If it is not possible to write automated tests, explain why and document how
  to manually test and verify the feature.
-->

### Is there any documentation that needs to be updated for this PR?

<!--
- If yes and the docs are in GitHub, include doc updates in the PR.
- If yes and the docs are not in GitHub (i.e. ADO wiki), include a link to the
  docs.
- If no, explain why (e.g. "tech debt cleanup, N/A").
-->

No

troy0820 commented on pkg/operator/controllers/node/node_controller_test.go:
```diff
@@ -0,0 +1,187 @@
+			want: false,
+		},
+		{
+			name: "isDraining false, node is not unschedulable",
```
nit: The double negative is creating confusion but it may be better to say 
```suggestion
			name: "isDraining false, node is unschedulable",
```

troy0820 commented on pkg/operator/controllers/node/node_controller_test.go:
```diff
@@ -0,0 +1,187 @@
+		want bool
+	}{
+		{
+			name: "isDraining false",
```
nit: name of test is misleading. Yes it is false but it is false because the node is unschedulable.  This seems to represent the same test on line 59

cadenmarchese replied:
This test covers us on this line: https://github.com/Azure/ARO-RP/blob/master/pkg/operator/controllers/node/node_controller.go#L187 and I am not sure how to rename it. Something like `isDraining false, no conditions are met`?

troy0820 replied:
If you were going to rename it to the above, I would create the test to check all the conditions that are not being met to satisfy the false on line 187.  This way you'll satisfy the test named `isDraining false, no conditions met`

troy0820 reviewed (dismissed):
Just a few nits.  You have coverage over all these functions but because you are doing a table test you could create a test that tests every condition in the `isDraining` function.  So the annotations to see if they are blank, etc.  That way the confidence of this function will increase as you tested that each condition can create the return value necessary.

m1kola reviewed (changes requested):
I think we need the following tests for this controller:

1. Test individual functions such as `isDraining` with all possible scenarios as Troy suggested.
2. Make `isDraining` mockable and test the controller via `Reconcile`. This way we don't have to meet all the conditions inside `isDraining` in test data and don't have to care about implementation of `isDraining`: we just make it returns either `true` or `false` and cover these code paths.

I'm happy for these two things to be submitted in a separate PRs, if you prefer. But we definitely need the second point - without it there is a lot of code (contriller's main logic) without coverage.

cadenmarchese commented:
@m1kola @troy0820 I agree we should test via Reconcile, and I had previously attempted to do it unsuccessfully, but I can work on it more. I have addressed the other comments and squashed commits, so we can either merge with followup issue to change to Reconcile, or else I will continue to work on it via this PR - whatever is preferred.

m1kola commented on pkg/operator/controllers/node/node_controller_test.go:
```diff
@@ -0,0 +1,353 @@
+			annotationKey:   "",
+			annotationValue: "",
+		},
+		{
+			name: "annotations are provided, set them",
+			node: &corev1.Node{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{"some-random-annotation": "is-set"},
+				},
+			},
+			annotationKey:   "some-random-annotation",
+			annotationValue: "is-set",
+		},
```
What ware we testing in this case? Try intentionally break the function you are testing to make sure that the test case is useful. Will this test case fail if we change `setAnnotation` to a no-op function like this?

```
func setAnnotation(m *metav1.ObjectMeta, k, v string) {
}
```

cadenmarchese replied:
After playing with it a bit, this test wasn't necessary. I have removed it.

m1kola reviewed (changes requested):
@cadenmarchese let's finish covering individual functions in this PR and cover `Reconcile` in a separate PR.

cadenmarchese commented on pkg/operator/controllers/node/node_controller_test.go:
```diff
@@ -0,0 +1,353 @@
+		annotationKey   string
+		annotationValue string
+	}{
+		{
+			name:            "no annotations set, set blank ones",
+			node:            &corev1.Node{},
+			annotationKey:   "",
+			annotationValue: "",
+		},
```
I've added values here.

cadenmarchese commented on pkg/operator/controllers/node/node_controller_test.go:
```diff
@@ -0,0 +1,353 @@
+				},
+			},
+			want: false,
+		},
```
Thanks for the tip, things look better now:
![nodecontrollerCoverage](https://user-images.githubusercontent.com/56140267/128768687-bd2f4c73-fbfd-4043-9372-ca186d23e706.png)

m1kola commented on pkg/operator/controllers/node/node_controller_test.go:
```diff
@@ -0,0 +1,312 @@
+			node:            &corev1.Node{},
+			annotationKey:   "foo",
+			annotationValue: "bar",
+		},
```
I think we need a case to make sure that we override an existing value.

m1kola reviewed (commented):
Looks good. Only one small suggestion for `TestSetAnnotation`.

m1kola reviewed (approved):
Looks good. We still need to cover main logic of this controller (`Reconcile` method). Let's do it in a separate PR.
